lib/src/protofier.dart

```diff
@@ -0,0 +1,274 @@
+// Copyright 2019 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:sass/sass.dart' as sass;
+
+import 'dispatcher.dart';
+import 'embedded_sass.pb.dart';
+import 'function_registry.dart';
+import 'host_callable.dart';
+import 'utils.dart';
+
+/// A class that converts Sass [sass.Value] objects into [Value] protobufs.
+///
+/// A given [Protofier] instance is valid only within the scope of a single
+/// custom function call.
+class Protofier {
+  /// The dispatcher, for invoking deprotofied [Value_HostFunction]s.
+  final Dispatcher _dispatcher;
+
+  /// The IDs of first-class functions.
+  final FunctionRegistry _functions;
+
+  /// The ID of the current compilation.
+  final int _compilationId;
+
+  /// Any argument lists transitively contained in [value].
+  ///
+  /// The IDs of the [Value_ArgumentList] protobufs are always one greater than
+  /// the index of the corresponding list in this array (since 0 is reserved for
+  /// argument lists created by the host).
+  final _argumentLists = <sass.SassArgumentList>[];
+
+  /// Creates a [Protofier] that's valid within the scope of a single custom
+  /// function call.
+  ///
+  /// The [functions] tracks the IDs of first-class functions so that the host
+  /// can pass them back to the compiler.
+  Protofier(this._dispatcher, this._functions, this._compilationId);
+
+  /// Converts [value] to its protocol buffer representation.
+  Value protofy(sass.Value value) {
+    var result = Value();
+    if (value is sass.SassString) {
+      result.string = Value_String()
+        ..text = value.text
+        ..quoted = value.hasQuotes;
+    } else if (value is sass.SassNumber) {
+      var number = Value_Number()..value = value.value * 1.0;
+      number.numerators.addAll(value.numeratorUnits);
+      number.denominators.addAll(value.denominatorUnits);
+      result.number = number;
+    } else if (value is sass.SassColor) {
+      // TODO(nweiz): If the color is represented as HSL internally, this coerces
+      // it to RGB. Is it worth providing some visibility into its internal
+      // representation so we can serialize without converting?
```

Comment:
I think this would be worth it. Would we also need to support HSL colors in the protocol?

Reply:
Yeah, that would be the implication. At that point we'd probably *also* need to support HWB. Given how much overhead is going to be taken up serializing and deserializing, I'm not sure it's worth it...

Reply:
I'd prefer the protocol to either (a) guarantee transparency on colour representation or (b) explicitly use some canonical rep [rgb?].  I have no idea about practical user-visible consequences of round-tripping hsl/rgb/hsl though.

eg. in my host Color implementation I added the 'preferred rep' method to know whether to gen `Value_RgbColor` or `Value_HslColor` but this is slightly pointless if the embedded side sometimes adds conversions: either the conversion is bad, so we shouldn't do it; or it's fine and so we may as well simplify current & future code (eg. hwb support) by doing it always and delete `Value_HslColor`.

(It feels to me that latency/cpu cost of doing additional conversions would be negligible compared to the cost of the whole protobuf IPC.)

Reply:
Awjin and I discussed this and agreed that, to match behavior with native Sass functions, the compiler should probably pass HSL and HWB colors as-is to the embedded host. I'll make that happen in a followup CL.